components/external_storage/export/src/export.rs

```diff
@@ -0,0 +1,363 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! To use External storage with protobufs as an application, import this module.
+//! external_storage contains the actual library code
+//! Cloud provider backends are under components/cloud
+use std::io::{self, Write};
+use std::path::Path;
+use std::sync::Arc;
+use std::time::Instant;
+
+#[cfg(feature = "cloud-aws")]
+pub use aws::{Config as S3Config, S3Storage};
+#[cfg(feature = "cloud-gcp")]
+pub use gcp::{Config as GCSConfig, GCSStorage};
+
+#[cfg(feature = "prost-codec")]
+pub use kvproto::backup::storage_backend::Backend;
+use kvproto::backup::CloudDynamic;
+#[cfg(feature = "protobuf-codec")]
+pub use kvproto::backup::StorageBackend_oneof_backend as Backend;
+#[cfg(any(feature = "cloud-gcp", feature = "cloud-aws"))]
+use kvproto::backup::{Gcs, S3};
+
+#[cfg(feature = "cloud-storage-dylib")]
+use external_storage::dylib_client;
+#[cfg(feature = "cloud-storage-dylib")]
+use crate::dylib;
+#[cfg(feature = "cloud-storage-grpc")]
+use external_storage::grpc_client;
+#[cfg(any(feature = "cloud-storage-dylib", feature = "cloud-storage-grpc"))]
+use cloud::blob::BlobConfig;
+use cloud::blob::BlobStorage;
+use encryption::DataKeyManager;
+use external_storage::record_storage_create;
+pub use external_storage::{
+    read_external_storage_into_file, ExternalStorage, LocalStorage, NoopStorage,
+};
+use futures_io::AsyncRead;
+use kvproto::backup::{Noop, StorageBackend};
+use tikv_util::stream::block_on_external_io;
+use tikv_util::time::Limiter;
+use tikv_util::warn;
+
+pub fn create_storage(storage_backend: &StorageBackend) -> io::Result<Box<dyn ExternalStorage>> {
+    if let Some(backend) = &storage_backend.backend {
+        create_backend(backend)
+    } else {
+        Err(bad_storage_backend(storage_backend))
+    }
+}
+
+// when the flag cloud-storage-dylib or cloud-storage-grpc is set create_storage is automatically wrapped with a client
+// This function is used by the library/server to avoid any wrapping
+pub fn create_storage_no_client(
+    storage_backend: &StorageBackend,
+) -> io::Result<Box<dyn ExternalStorage>> {
+    if let Some(backend) = &storage_backend.backend {
+        create_backend_inner(backend)
+    } else {
+        Err(bad_storage_backend(storage_backend))
+    }
+}
+
+fn bad_storage_backend(storage_backend: &StorageBackend) -> io::Error {
+    io::Error::new(
+        io::ErrorKind::NotFound,
+        format!("bad storage backend {:?}", storage_backend),
+    )
+}
+
+fn bad_backend(backend: Backend) -> io::Error {
+    let storage_backend = StorageBackend { backend: Some(backend), ..Default::default() };
+    bad_storage_backend(&storage_backend)
+}
+
+#[cfg(any(feature = "cloud-gcp", feature = "cloud-aws"))]
+fn blob_store(store: Box<dyn BlobStorage>) -> Box<dyn ExternalStorage> {
```

Comment:
i'd suggest turning `BlobStore` to a generic type to avoid the doubling boxing.

```rust
struct BlobStore<B>(B);
impl<B: BlobStorage> ExternalStorage for BlobStore<B> { ... }
```

the `create_backend_inner` can use

```rust
Box::new(BlobStore(S3Storage::from_input(config.clone())?)))
```

Reply:
okay. I have found it really easy to use Box but I will try to start using generics where possible.